### DISCUSSION TITLE
Support of different RNG signatures in unified pipelines

### DISCUSSION THREAD

**Discussion by PgLoLo** [google/flax#1513] [Q&A]
I don't understand how to work with rng keys for linen Modules. Consider some simple training pipeline, that on each step runs the model, computes loss and gradients and updates parameters. We want to implement this pipeline in a model-agnostic way. But different neural network architectures may require different rng signatures: ResNets usually does not require rng at all, VGG want a dictionary `{'dropout': rng}`, some other model may require a much more complex rng structure.  

There is no big deal to make each step in the training loop require the simple rng key, but how can I make this key automatically split into dictionary structure with the signature, required by the model?

**Comment by devjeetr** [comment]
I'm not quite sure if this is the most optimal approach, but what I do is define a thin layer around `linen` modules that contains auxiliary model specific logic for training/inference. This includes an `init` function that initializes params and rngs. It looks like this:
```python
# needs to be defined for each model
def init_model(rng, model_cls, cfg) -> Tuple[FrozenVariableDict, Dict[str, Any]] :
    dropout_rng, k1_rng, k2_rng = jax.random.split(rng, 3)
    model_rngs = {
        'dropout': dropout_rng,
        'k1': k1_rng,
        'k2': k2_rng
    }
    
    params = model(cfg).init(...)['params']

    return params, model_rngs
```
You can then put the initialized rngs into `TrainState`, and modify replication/folding of the rngs by tree_mapping over the rng pytree:

```python
from flax.training.train_state import TrainState
from flax.training.common_utils import shard_prng_key

# model agnostic
class ModifiedTrainState(TrainState):
    def replicate(self):
        return jax_utils.replicate(self).replace(
            rngs=jax.tree_map(shard_prng_key, self.rngs)
        )


def train_step(state, ...):
    
    rngs = jax.tree_map(lambda rng: jax.random.fold_in(rng, state.step), state.rngs) 

    ... perform train step using folded rngs...

    state = state.replace(rngs=rngs)
    
    return state, ...
```

Would love to hear if there are other patterns people use to deal with this!

**Reply by PgLoLo** [reply]
Thank you for the answer. So basically, what you do here is provide the rng dictionary template and then split rng key into this template, right? 

I wonder, why flax does not do it automatically while I call the `apply` method. If I provide a simple rng key to the apply method, I would get an error "this module requires that rng key in `rngs` dictionary, please fix". Why in that case flax does not automatically recursively split my `rng` for submodules if it already knows the required structure?

Because for situations like encoder-decoder modules, that can be parameterized with different encoders and different encoders the logic suggested by you (with providing rng dictionary template) would require encapsulating it into the module interface itself. Which is, well, very doable while you are working on your own codebases only, but really complicates things in situations when neural network architectures comes from other people libraries.

**Comment by manifest** [comment]
Similar to Devjeet, I've just written a [function](https://github.com/manifest/flax-extra/blob/v0.0.1/src/flax_extra/random.py#L20-L33) that consumes a list of collection labels and produces labeled random number generator keys. [Here](https://flax-extra.readthedocs.io/en/latest/notebooks/training/) is how it is used in the training loop.

> I wonder, why flax does not do it automatically while I call the `apply` method. If I provide a simple rng key to the apply method, I would get an error "this module requires that rng key in `rngs` dictionary, please fix". Why in that case flax does not automatically recursively split my `rng` for submodules if it already knows the required structure?

I have the same question. I wonder if that is really a problem to split keys for all the random collections even though they wouldn't be used. If splitting all the keys is not an option, maybe models should define labels required for the inference and/or training (e.g. something like `model.random_collections() -> [str]`). That should cover most of the use cases.
